## Lession 1 : First Steps with Python and Jupyter

This tutorial series is a beginner-friendly introduction to programming and data analysis using the Python programming language. These tutorials take a practical and coding-focused approach. The best way to learn the material is to execute the code and experiment with it yourself. Check out the full series here: 

1. [First Steps with Python and Jupyter](https://jovian.ai/aakashns/first-steps-with-python)
2. [A Quick Tour of Variables and Data Types](https://jovian.ai/aakashns/python-variables-and-data-types)
3. [Branching using Conditional Statements and Loops](https://jovian.ai/aakashns/python-branching-and-loops)
4. [Writing Reusable Code Using Functions](https://jovian.ai/aakashns/python-functions-and-scope)
5. [Reading from and Writing to Files](https://jovian.ai/aakashns/python-os-and-filesystem)
6. [Numerical Computing with Python and Numpy](https://jovian.ai/aakashns/python-numerical-computing-with-numpy)
7. [Analyzing Tabular Data using Pandas](https://jovian.ai/aakashns/python-pandas-data-analysis)
8. [Data Visualization using Matplotlib & Seaborn](https://jovian.ai/aakashns/python-matplotlib-data-visualization)
9. [Exploratory Data Analysis - A Case Study](https://jovian.ai/aakashns/python-eda-stackoverflow-survey)

This tutorial covers the following topics:
* Performing arithmetic operations using Python
* Solving multi-step problems using variables
* Evaluating conditions using Python
* Combining conditions with logical operators
* Adding text styles using Markdown

## How to run the code

This tutorial is an executable [Jupyter notebook](https://jupyter.org/) rebuilt by [IACentre](https://www.linkedin.com/in/iacentre/). You can *run* this tutorial and experiment with the code examples in a couple of ways: using free online resources (recommended) or on your computer.

## Running on your computer locally

To run the code on your computer locally, you'll need to set up [Python](https://www.python.org/), download the notebook and install the required libraries. We recommend using the [Conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/) distribution of Python. Click the **Run button** at the top of this page, select the Run Locally option, and follow the instructions.

>**Jupyter Notebooks**: This tutorial is a [Jupyter notebook] - a document made of cells. Each cell can contain code written in Python or explanations in plain English. You can execute code cells and view the results, e.g., numbers, messages, graphs, tables, files, etc. instantly within the notebook. Jupyter is a powerful platform for experimentation and analysis. Don't be afraid to mess around with the code & break things - you'll learn a lot by encountering and fixing errors. You can use the "Kernel > Restart & Clear Output" menu option to clear all outputs and start again from the top.

## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter.
>Working with cells: To create a new cell within Jupyter, you can select "Insert > Insert Cell Below" from the menu bar or just press the "+" button on the toolbar. You can also use the keyboard shortcut Esc+B to create a new cell. Once a cell is created, click on it to select it. You can then change the cell type to code or markdown (text) using the "Cell > Cell Type" menu option. You can also use the keyboard shortcuts Esc+Y and Esc+M. Double-click a cell to edit the content within the cell. To apply your changes and run a cell, use the "Cell > Run Cells" menu option or click the "Run" button on the toolbar or just use the keyboard shortcut Shift+Enter. You can see a full list of keyboard shortcuts using the "Help > Keyboard Shortcuts" menu option.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the modified cells again to see updated results. Can you guess what the //, %, and ** operators are used for?

In [1]:
2 + 3 + 9

14

In [2]:
99 - 73

26

In [3]:
23.54 * -1432

-33709.28

In [4]:
100 / 7

14.285714285714286

In [5]:
100 // 7 #Floor Division

14

In [6]:
11 % 7 #Reminder / Modulus

4

In [7]:
5 ** 3 #Power of 3 i.e. eponent

125

Try solving some simple problems from this page: [Worksheet on Word Problems on Four Operations](https://www.math-only-math.com/worksheet-on-word-problems-on-four-operations.html) .

## Solving multi-step problems using variables

Let's try solving the following word problem using Python:
>A grocery store sells a bag of ice for $1.25 and makes a 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided and gradually convert the word problem into a mathematical expression that can be evaluated using Python.

*Cost of ice bag ($)*= 1.25

*Profit margin* = 20% = .2

*Profit per bag ($)* = profit margin * cost of ice bag = .2 * 1.25

*No. of bags* = 500

*Total profit* = no. of bags * profit per bag = 500 * (.2 * 1.25)

In [8]:
500 * (0.2 * 1.25)

125.0

Thus, the grocery store makes a total profit of $125. While this is a reasonable way to solve a problem, it's not entirely clear by looking at the code cell what the numbers represent. We can give names to each of the numbers by creating Python variables.

>Variables: While working with a programming language such as Python, information is stored in *variables*. You can think of variables as containers for storing data. The data stored within a variable is called its *value*.

In [9]:
cost_of_ice_bag = 1.25

In [10]:
profit_margin = .2

In [11]:
number_of_bags = 500

The variables `cost_of_ice_bag`, `profit_margin`, and `number_of_bags` now contain the information provided in the word problem. We can check the value of a variable by typing its name into a cell. We can combine variables using arithmetic operations to create other variables.

>Code completion: While typing the name of an existing variable in a code cell within Jupyter, just type the first few characters and press the Tab key to autocomplete the variable's name. Try typing pro in a code cell below and press Tab to autocomplete to `profit_margin`.

In [12]:
profit_margin

0.2

In [13]:
profit_per_bag = cost_of_ice_bag * profit_margin

In [14]:
profit_per_bag

0.25

In [15]:
total_profit = number_of_bags * profit_per_bag

In [16]:
total_profit

125.0

Let's display the result of the word problem using a friendly message. We can do this using the print function.
>**Functions**: A function is a reusable set of instructions. It takes one or more inputs, performs certain operations, and often returns an output. Python provides many in-built functions like print and also allows us to define our own functions.

In [17]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


>**print**: The print function is used to display information. It takes one or more inputs, which can be text (within quotes, e.g., "this is some text"), numbers, variables, mathematical expressions, etc. We'll learn more about variables & functions in the next tutorial.

Creating a code cell for each variable or mathematical operation can get tedious. Fortunately, Jupyter allows you to write multiple lines of code within a single code cell. The result of the last line of code within the cell is displayed as the output.

Let's rewrite the solution to our word problem within a single cell.

In [18]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag
net_profit = (cost_of_ice_bag * (1 + profit_margin) * number_of_bags) - (number_of_bags * cost_of_ice_bag)

# Display the result
print("The grocery store makes a total profit of $", total_profit)
print("The grocery store makes a net profit of $", net_profit)

The grocery store makes a total profit of $ 125.0
The grocery store makes a net profit of $ 125.0


Note that we're using the # character to add comments within our code.

>**Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines.

Inline and single-line comments start with #, whereas multi-line comments begin and end with three quotes, i.e. """. Here are some examples of code comments:

In [19]:
my_favorite_number = 1 # an inline comment

In [20]:
# This comment gets its own line
my_least_favorite_number = 3

In [21]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5

In [22]:
a_neutral_number

5

**EXERCISE**: A travel company wants to fly a plane to the Bahamas. Flying the plane costs 5000 dollars. So far, 29 people have signed up for the trip. If the company charges 200 dollars per ticket, what is the profit made by the company? Create variables for each numeric quantity and use appropriate arithmetic operations.

In [23]:
# Store input data in variables
flying_cost = 5000
number_clients = 29
price_ticket = 200

# Perform the required calculations
total_revenue = number_clients * price_ticket
total_profit = total_revenue - flying_cost 

# Display the result
print("The travle coompany makes a total profit of $", total_profit)

The travle coompany makes a total profit of $ 800


## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

|   Operator   |   Descriptio   |
|   :-:   |   -:   |
|  ` ==`	|   Check if operands are equal  |
|   `!=`	|   Check if operands are not equal   |
|   `>`	|   Check if operand is greater than right operand   |
|   `<`	|   Check if operand is greater than left operand  |
|   `>=`	|   Check if operand is greater than or equal to right operand  
|   `<=`	|   Check if operands are equal is less than or equal to right operand   |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.


In [24]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [25]:
# Equality check - True
my_favorite_number == 1

True

In [26]:
# Equality check - False
my_favorite_number == my_least_favorite_number
False

False

In [27]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [28]:
# Not equal check - False
a_neutral_number != 3

False

In [29]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [30]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [31]:
# Less than check - True
my_favorite_number < 10

True

In [32]:
# Less than check - False
my_least_favorite_number < my_favorite_number

False

In [33]:
# Less than or equal check - True
3 + 6 <= 9

True

In [34]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

False

In [35]:
#Just like arithmetic operations, the result of a comparison operation can also be stored in a variable.

cost_of_ice_bag = 1.25 #You can change the value till 10 , the statement will turn to True
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

|   `a`   |   `b`  | `a and b` |
|   :-:   |   :-:  |    :-:    |
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |

In [36]:
my_favorite_number

1

In [37]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [38]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [39]:
my_favorite_number > 0 and my_favorite_number >= 3

False

In [40]:
True and False

False

In [41]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|:-:|:-:|:-:|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [42]:
a_neutral_number = 3

In [43]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [44]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [45]:
my_favorite_number < 0 or True

True

In [46]:
False or False

False

The `not` operator returns `False` if a condition is `True` and `True` if the condition is `False`.

In [47]:
not a_neutral_number == 3

False

In [48]:
not my_favorite_number < 0

True

In [49]:
not False

True

In [50]:
not True

False

Logical operators can be combined to form complex conditions. Use round brackets or parentheses `(` and `)` to indicate the order in which logical operators should be applied.

In [51]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [52]:
not (True and 0 < 1) or (False and True)

False

In [53]:
#If parentheses are not used, logical operators are applied from left to right.
not True and 0 < 1 or False and True

False

Experiment with arithmetic, conditional and logical operators in Python using the interactive nature of Jupyter notebooks. We will learn more about variables and functions in future tutorials.

## Adding text styles using Markdown
Adding explanations using text cells (like this one) is a great way to make your notebook informative for other readers. It is also useful if you need to refer back to it in the future. You can double click on a text cell within Jupyter to edit it. In the edit mode, you'll notice that the text looks slightly different (for instance, the heading has a ## prefix. This text is written using Markdown, a simple way to add styles to your text. Execute this cell to see the output without the special characters. You can switch back and forth between the source and the output to apply a specific style.

For instance, you can use one or more # characters at the start of a line to create headers of different sizes:

# Header 1

## Header 2

### Header 3 

#### Header 4

To create a bulleted or numbered list, simply start a line with * or 1..

A bulleted list:

* Item 1
* Item 2
* Item 3

A numbered list:

1. Apple
2. Banana
3. Pineapple

A bold text with `**`:
** Jupyter Notebook

A Italic Test with `*text*`
*Jupyter Notebook*

Another really nice feature of Markdown is the ability to include blocks of code. Note that code blocks inside Markdown cells cannot be executed.

#### Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

#### Display the result
print("The grocery store makes a total profit of $", total_profit)

You can learn the full syntax of Markdown here: https://learnxinyminutes.com/docs/markdown/

### Further Reading and References

Following are some resources where you can learn about more arithmetic, conditional and logical operations in Python:

* Python Tutorial at W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html
* Python official documentation: https://docs.python.org/3/tutorial/index.html

Now that you have taken your first steps with Python, you are ready to move on to the next tutorial: ["A Quick Tour of Variables and Data Types in Python"](https://jovian.ml/aakashns/python-variables-and-data-types).